### PR TITLE
Recognize passthrough for outbound throttling.

### PR DESCRIPTION
**Pull request by riyafa**

### CODE REVIEW

File: stdlib/http/src/main/java/org/ballerinalang/net/http/actions/httpclient/AbstractHTTPAction.java [hunk 1]
```diff
@@ -358,10 +357,20 @@ private void send(DataContext dataContext, HttpCarbonMessage outboundRequestMsg,
...
+            if (requestStruct != null) {
+                BMap<String, BValue> entityStruct = extractEntity(requestStruct);
+                if (entityStruct != null) {
+                    if (boundaryString != null) {
+                        serializeMultiparts(entityStruct, messageOutputStream, boundaryString);
+                    } else {
+                        serializeDataSource(entityStruct, messageOutputStream);
+                    }
+                } else {
+                    //This is not really a passthrough but is the case when the body is null/not built.
```
Comment (arukshani) [inline stdlib/http/src/main/java/org/ballerinalang/net/http/actions/httpclient/AbstractHTTPAction.java]: This comment is a bit ambiguous. Can we rephrase it to give the exact meaning as to why it is set as passthrough? Something along the line that it reaches here when it is a genuine passthrough (the body has not been read) or the entity body is actually null. Since there's no way to distinguish between the two from ballerina side, it is always set as true but it does no harm and transport side will interpret this value only when there's an unread body in carbon message.

**Comment by Shan1024** [comment]
@arukshani - Shall we resolve the conflicts?

**Comment by Shan1024** [comment]
@riyafa - Shall we resolve the conflicts?

File: stdlib/http/src/main/java/org/ballerinalang/net/http/actions/httpclient/AbstractHTTPAction.java [hunk 2]
```diff
@@ -393,29 +405,21 @@ private HttpMessageDataStreamer getHttpMessageDataStreamer(HttpCarbonMessage out
...
-                                             entityStruct);
-            } else { //If the content is in a byte channel
-                serializeDataSource(context, messageOutputStream);
-            }
+    private void serializeMultiparts(BMap<String, BValue> entityStruct, OutputStream messageOutputStream,
+                                     String boundaryString) throws IOException {
+        BRefValueArray bodyParts = EntityBodyHandler.getBodyPartArray(entityStruct);
+        if (bodyParts != null && bodyParts.size() > 0) {
+            serializeMultipartDataSource(messageOutputStream, boundaryString,
+                                         entityStruct);
```
Comment (Shan1024) [inline stdlib/http/src/main/java/org/ballerinalang/net/http/actions/httpclient/AbstractHTTPAction.java]: Can merge with the line above.

File: stdlib/http/src/main/java/org/ballerinalang/net/http/actions/httpclient/AbstractHTTPAction.java [hunk 3]
```diff
@@ -430,23 +434,16 @@ private void serializeMultipartDataSource(OutputStream messageOutputStream,
...
-            if (messageDataSource != null) {
-                HttpUtil.serializeDataSource(messageDataSource, entityStruct, messageOutputStream);
+    private void serializeDataSource(BMap<String, BValue> entityStruct, OutputStream messageOutputStream)
+            throws IOException {
+        BValue messageDataSource = EntityBodyHandler.getMessageDataSource(entityStruct);
+        if (messageDataSource != null) {
+            HttpUtil.serializeDataSource(messageDataSource, entityStruct, messageOutputStream);
+            HttpUtil.closeMessageOutputStream(messageOutputStream);
+        } else { //When the entity body is a byte channel and when it is not null
+            if (EntityBodyHandler.getByteChannel(entityStruct) != null) {
```
Comment (Shan1024) [inline stdlib/http/src/main/java/org/ballerinalang/net/http/actions/httpclient/AbstractHTTPAction.java]: Can merge the `else` and `if` and have a single `else if` for this?

**Comment by anupama-pathirage** [comment]
@Riyafa Can you please resolve conflicts in this PR?

File: tests/ballerina-integration-test/src/test/java/org/ballerinalang/test/data/streaming/TableToXMLStreamingTestCase.java [hunk 4]
```diff
@@ -79,6 +81,15 @@ public void testStreamingLargeXML() throws Exception {
         Assert.assertEquals(Integer.parseInt(response.getData()), 211288909);
     }
 
+    @Test(description = "Tests streaming a large amount of data from a table, converted to XML")
```
Comment (anupama-pathirage) [inline tests/ballerina-integration-test/src/test/java/org/ballerinalang/test/data/streaming/TableToXMLStreamingTestCase.java]: Shall we add proper description here? Seems a copied content.
